GitUI/UserControls/NativeTreeView.cs

```diff
@@ -8,4 +12,100 @@ protected override void CreateHandle()
             NativeMethods.SetWindowTheme(Handle, "explorer", null);
         }
     }
+
+    // Add double-click related behaviour to NativeTreeView
+    public class NativeTreeViewDoubleClickDecorator
+    {
+        private NativeTreeView _treeView;
+
+        private DateTime _lastMouseDown = DateTime.MinValue;
+        private int _mouseClicks = 0;
+
+        private DateTime _lastDoubleClick = DateTime.MinValue;
+
+        // Invoked just before an inner tree node is expanded/collapsed. Set CancelEventHandler.Cancel to true
+        // to cancel the expand/collapse of the current node.
+        public event CancelEventHandler BeforeDoubleClickExpandCollapse;
+
+        public NativeTreeViewDoubleClickDecorator(NativeTreeView treeView)
+        {
+            _treeView = treeView;
+            _treeView.Disposed += OnDisposed;
+            _treeView.BeforeCollapse += OnBeforeCollapse;
+            _treeView.BeforeExpand += OnBeforeExpand;
+            _treeView.MouseDown += OnMouseDown;
+        }
+
+        private void OnDisposed(object sender, EventArgs e)
+        {
+            _treeView.BeforeCollapse -= OnBeforeCollapse;
+            _treeView.BeforeExpand -= OnBeforeExpand;
+            _treeView.MouseDown -= OnMouseDown;
+            _treeView = null;
+        }
+
+        private void OnBeforeCollapse(object sender, TreeViewCancelEventArgs e)
+        {
+            HandleBeforeDoubleClickExpandCollapse(sender, e);
+        }
+
+        private void OnBeforeExpand(object sender, TreeViewCancelEventArgs e)
+        {
+            HandleBeforeDoubleClickExpandCollapse(sender, e);
+        }
+
+        private void HandleBeforeDoubleClickExpandCollapse(object sender, TreeViewCancelEventArgs e)
+        {
+            // Once detected, reset double-clicked state so that if BeforeDoubleClickExpandCollapse does
+            // anything that collapses/expands nodes on the tree, this function won't erroneously resend
+            // this event.
+            bool doubleClicked = DoubleClicked();
+            ResetDoubleClick();
```

Comment:
Should we reset only if we double clicked?

Reply:
No, we need to always reset once this handler is called, which I attempted to explain in the comment above.